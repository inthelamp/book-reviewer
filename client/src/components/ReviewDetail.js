import { useEffect, useState } from 'react'
import axios from 'axios'
import PropTypes from 'prop-types'
import { Container, ButtonGroup, Row, Col, Form, Button } from 'react-bootstrap'
import ReadOnlyText from './ReadOnlyText'
import Message from '../components/Message'

/**
 * @typedef {Object} Review
 * @property {string} reviewId - Review id
 * @property {subject} subject - Review subject
 * @property {bool} isOwner - presenting if the user is the owner of the review or not
 * @property {string} status - Review status
 */ 
/**
 * Displaying review content and updating review information
 * @param {string} userId - User id
 * @param {bool} isSignedIn - Presenting if the user is signed in or not
 * @param {Review} review - Review object
 */
const ReviewDetail = ( { userId, isSignedIn, review } ) => { 
    const [detail, setDetail] = useState()
    const [subject, setSubject] = useState('')
    const [bookTitle, setBookTitle] = useState('')
    const [isbn, setISBN] = useState('')
    const [authors, setAuthors] = useState({
            firstAuthor: '',
            secondAuthor: '',
            thirdAuthor: ''
    }) 
    const [message, setMessage] = useState('')
    const [messageStyle, setMessageStyle] = useState('')

    // Getting a review  
    useEffect(() => {  
        if (review.reviewId) {
            // Sending server a request to get others' reviews
            axios
            .get(process.env.REACT_APP_SERVER_BASE_URL + '/reviews/reviews/' + review.reviewId)
            .then((response) => {    
                setDetail(response.data.detail)         
                console.log(response.data.message)                  
            })
            .catch ((error) => {   
                if (error.response) {
                    console.log(error.response.data.message)              
                } else {
                    console.log('Error', error.message)
                }
            })   
        }
    }, [review.reviewId]); 

    const containerStyle = {
        paddingLeft: 0,
        paddingRight: 0,
    }

    const updateAuthor = (e) => {     
        e.preventDefault()
        
        setAuthors({
            ...authors,
            [e.target.name]: e.target.value
        });
    }

    const onSubmit = (e) => {
        e.preventDefault()

        setMessage('Hi, it needs further development!')
        setMessageStyle('success-message')       
        alert('Hi, it needs further development!');
    }

    const onEditThisReview = (e) => {
        e.preventDefault()

        setMessage('Hi, it needs further development!')
        setMessageStyle('success-message')       
        alert('Hi, it needs further development!');
    }

    const onMakeNewPost = (e) => {
        e.preventDefault()

        setMessage('Hi, it needs further development!')
        setMessageStyle('success-message')       
        alert('Hi, it needs further development!');
        console.log(userId)
    }

    return (
        <>
             <ButtonGroup className='mb-1 float-sm-end'>
                { isSignedIn ?  
                    //  Displaying this edit button if the user is signed in
                    <Button variant='secondary' onClick={onEditThisReview} >Edit this review</Button>   
                    : null 
                }
                <Button variant='secondary' onClick={onMakeNewPost}>Post new review</Button>
            </ButtonGroup>    

            {detail ? <ReadOnlyText content={detail.content} /> : null }

            {/* Allowed for reviews posted by the use signed in */}
            {review.isOwner ? 
                <Form className='mt-2' onSubmit={onSubmit}>    
                    <Form.Group className='mb-3' size='lg' controlId='subject'>
                        <Form.Label>Subject</Form.Label>
                        <Form.Control
                        type='text'
                        value={subject}
                        onChange={(e) => setSubject(e.currentTarget.value)}
                        />      
                    </Form.Group>
                    <Form.Group className='mb-3' size='lg' controlId='bookTitle'>
                    <Container style={containerStyle}> 
                        <Row>
                            <Col xs={10}>
                                <Form.Label>Book Title</Form.Label>
                                <Form.Control
                                    type='text'
                                    value={bookTitle}
                                    onChange={(e) => setBookTitle(e.currentTarget.value)}
                                />     
                            </Col>
                            <Col xs={2}>
                                <Form.Label>Book ISBN-13</Form.Label>
                                <Form.Control
                                    type='text'
                                    value={isbn}
                                    onChange={(e) => setISBN(e.currentTarget.value)}    
                                />    
                            </Col>
                        </Row>
                    </Container>    
                    </Form.Group>          
                    <Form.Group className='mb-3' size='lg' controlId='authors'>
                    <Container style={containerStyle}>
                        <Row>
                            <Col>
                                <Form.Label>First Author</Form.Label>
                                <Form.Control
                                    type='text'
                                    name='firstAuthor'
                                    value={authors.firstAuthor}
                                    onChange={updateAuthor}
                                />     
                            </Col>
                            <Col>
                                <Form.Label>Second Author</Form.Label>
                                <Form.Control
                                    type='text'
                                    name='secondAuthor'                               
                                    value={authors.secondAuthor}
                                    onChange={updateAuthor}
                                />    
                            </Col>
                            <Col>
                                <Form.Label>Third Author</Form.Label>
                                <Form.Control
                                    type='text'
                                    name='thirdAuthor'                                     
                                    value={authors.thirdAuthor}
                                    onChange={updateAuthor}
                                />    
                            </Col>
                        </Row>
                    </Container>
                    </Form.Group>                    
                    <div className='App-Buttons'>
                        <Button size='lg' type='submit'>Update</Button>     
                        <Message message={message} messageStyle={messageStyle}/>                         
                    </div>      
                </Form>   
                :
                null 
            }
        </>
    )
}

ReviewDetail.propTypes = {
    userId: PropTypes.string,
    isSignedIn: PropTypes.bool.isRequired,
    review: PropTypes.shape({
        reviewId: PropTypes.string,
        subject: PropTypes.string,
        isOwner: PropTypes.bool,
        status: PropTypes.string        
    })
}

export default ReviewDetail
