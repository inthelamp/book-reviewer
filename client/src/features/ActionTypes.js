const ActionTypes = {
	SIGNED_IN: 'signedIn',
	SIGNED_OUT: 'signedOut'
}

export default ActionTypes