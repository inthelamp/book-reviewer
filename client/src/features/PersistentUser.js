import jwt_decode from 'jwt-decode'

export const getToken = () => {
    const name = 'token='
    const decodedCookie = decodeURIComponent(document.cookie)
    const cookieAttributes = decodedCookie.split(';')

    for(let i = 0; i <cookieAttributes.length; i++) {
        let ca = cookieAttributes[i]
        while (ca.charAt(0) === ' ') {
            ca = ca.substring(1)
        }
        if (ca.indexOf(name) === 0) {
            return ca.substring(name.length, ca.length)
        }
    }
    return ''
}

export const isTokenExpried = () => {
    const token = getToken()

    if (token) {      
        try {
            const decoded = jwt_decode(token)     
            const { exp } = decoded

            if (Date.now() < exp * 1000) { 
                return false
            } 
        } catch (error) {
            console.log(error)
        }
    }   

    return true
}

export const persistentUser = () => {
    const token = getToken()

    if (token) {      
        try {
            const decoded = jwt_decode(token)     
            const { name, role, exp } = decoded

            if (Date.now() < exp * 1000) { 
                return {
                    name,
                    role,
                    "isSignedIn": true
                }
            } else {
                return {
                    "isTokenExpried": true
                }
            }
        } catch (error) {
            console.log(error)
        }
    }

    return null
}