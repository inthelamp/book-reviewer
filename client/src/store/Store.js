import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'


//const rootReducer = combineReducers({User: UserReducer})
const Store = createStore(UserReducer)

export default Store