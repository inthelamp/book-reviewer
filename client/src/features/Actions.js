import { User, TextEditor, Review } from './ActionTypes'

/**
 * SignedIn action
 * @param {string} id - user id
 * @param {string} name - user name
 * @param {string} role - user role
 */
export const SignedIn = (id, name, role) => ({
    type: User.SIGNED_IN,
    payload: {
        id,
        name, 
        role
    }
})

// SignedOut action
export const SignedOut = () => ({
    type: User.SIGNED_OUT
})

// ChangedModeToBrowse action
export const ChangedModeToBrowse = (review) => ({
    type: TextEditor.CHANGED_MODE_TO_BROWSE,
    payload: {
        review
    }
})

// ChangedModeToEdit action
export const ChangedModeToEdit = (review) => ({
    type: TextEditor.CHANGED_MODE_TO_EDIT,
    payload: {
        review
    }
})

// PostedReview action
export const PostedReview = () => ({
    type: Review.POSTED_REVIEW
})

// ListedReview action
export const ListedReview = () => ({
    type: Review.LISTED_REVIEW
})