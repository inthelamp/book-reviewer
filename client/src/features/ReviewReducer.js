import { Review } from './ActionTypes'

// Review reducer
const ReviewReducer = (state = [], action) => {
    switch(action.type) {
        case Review.POSTED_REVIEW:
            return {
                ...state, 
                status: Review.POSTED_REVIEW,         
            }
            
        case Review.LISTED_REVIEW:
            return {
                ...state, 
                status: Review.LISTED_REVIEW,         
            }
            
        default:
            return state    
    }
}

export default ReviewReducer
