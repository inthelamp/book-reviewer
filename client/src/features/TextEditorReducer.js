import { TextEditor } from './ActionTypes'

// TextEditor reducer
const TextEditorReducer = (state = [], action) => {
    switch(action.type) {
        case TextEditor.CHANGED_MODE_TO_BROWSE:
            return {
                ...state, 
                review: action.payload.review,   
                mode: TextEditor.CHANGED_MODE_TO_BROWSE,           
            }
            
        case TextEditor.CHANGED_MODE_TO_EDIT:
            return {
                ...state, 
                review: action.payload.review,                
                mode: TextEditor.CHANGED_MODE_TO_EDIT
            }
            
        default:
            return state    
    }
}

export default TextEditorReducer
