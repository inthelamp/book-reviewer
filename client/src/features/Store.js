import { createStore, combineReducers } from 'redux'
import UserReducer from './UserReducer'
import TextEditorReducer from './TextEditorReducer'

// Creating store
const rootReducer = combineReducers({ User: UserReducer, TextEditor: TextEditorReducer})
const Store = createStore(
        rootReducer,
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() 
)

export default Store