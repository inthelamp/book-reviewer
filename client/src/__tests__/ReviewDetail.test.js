import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import ReviewDetail from '../components/ReviewDetail'
import { Review } from '../features/ActionTypes'
import Store from '../features/Store'


describe('Component: ReviewDetail', () => {
    it('Component: ReviewDetail rendering correctly', () => {  

        const userId = 'aaaaa'
        const isSignedIn = true
        const review = {
            reviewId: 'review-id-xxxxxx',
            subject: 'test subject',
            isOwner: true,
            status: Review.POSTED_REVIEW
        }

        const tree = renderer
          .create(<Provider store={Store}><ReviewDetail userId={userId} isSignedIn={isSignedIn} review={review} /></Provider>)
          .toJSON();
        expect(tree).toMatchSnapshot()
    })
})