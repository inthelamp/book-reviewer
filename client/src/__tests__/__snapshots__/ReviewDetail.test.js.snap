// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ReviewDetail Component: ReviewDetail rendering correctly 1`] = `
Array [
  <div
    className="mb-1 float-sm-end btn-group"
    role="group"
  >
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Edit this review
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Post new review
    </button>
  </div>,
  <form
    className="mt-2"
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="subject"
      >
        Subject
      </label>
      <input
        className="form-control"
        id="subject"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
      size="lg"
    >
      <div
        className="container"
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-10"
          >
            <label
              className="form-label"
              htmlFor="bookTitle"
            >
              Book Title
            </label>
            <input
              className="form-control"
              id="bookTitle"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="col-2"
          >
            <label
              className="form-label"
              htmlFor="bookTitle"
            >
              Book ISBN-13
            </label>
            <input
              className="form-control"
              id="bookTitle"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="mb-3"
      size="lg"
    >
      <div
        className="container"
        style={
          Object {
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="authors"
            >
              First Author
            </label>
            <input
              className="form-control"
              id="authors"
              name="firstAuthor"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="authors"
            >
              Second Author
            </label>
            <input
              className="form-control"
              id="authors"
              name="secondAuthor"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="col"
          >
            <label
              className="form-label"
              htmlFor="authors"
            >
              Third Author
            </label>
            <input
              className="form-control"
              id="authors"
              name="thirdAuthor"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="App-Buttons"
    >
      <button
        className="btn btn-primary btn-lg"
        disabled={false}
        type="submit"
      >
        Update
      </button>
      
    </div>
  </form>,
]
`;
